test/Streamly/Test/FileSystem/Event/Common.hs

```diff
@@ -0,0 +1,294 @@
+                -- they occur before the watch is started.  Even if we add a
+                -- delay here.
+                startWatchAndCheck root target sync
+
+    startWatchAndCheck root target sync = do
+            pre root
+            -- XXX The events from pre ops also seem to be bundled with the
+            -- events occurred after the watch is started.
+            let check = checker root target sync expected
+                fsOps = Stream.fromEffect $ runFSOps root sync
+            Stream.drain $ Stream.fromEffect check `Stream.parallelFst` fsOps
+
+    runFSOps fp sync = do
+        -- We put the MVar before the event watcher starts to run but that does
+        -- not ensure that the event watcher has actually started. So we need a
+        -- delay as well. Do we?
+        takeMVar sync >> threadDelay 200000
+        ops fp
+        threadDelay 10000000
+        error "Time out occurred before event watcher could terminate"
+
+-------------------------------------------------------------------------------
+-- Test descriptions
+-------------------------------------------------------------------------------
+
+-- XXX Cover more root level event tests
+-- XXX Tests for root as a regular file instead of dir?
+
+rootTests :: String -> [TestDesc]
+rootTests desc =
+    [
+    -- The watch root create event always seems to come even though the root is
+    -- created before the watch is started.
+      ( desc <> ", isCreated"
+      , const (return ())
+      , const (return ())
+      , [ ("", Event.isCreated) ]
+      )
+    ]
+
+dirEvent, fileEvent :: (Event -> Bool) -> Event -> Bool
+
+#if defined(FILESYSTEM_EVENT_DARWIN)
+dirEvent f ev = Event.isDir ev && f ev
+fileEvent f ev = Event.isFile ev && f ev
+#else
+dirEvent = id
+fileEvent = id
+#endif
+
+-- Tests only for regular dir roots, not for symlink roots
+-- For symlink roots, the deletion event does not come for the symlink.
+rootTestsRegular :: String -> [TestDesc]
+rootTestsRegular desc =
+    [
+    -- , let p = "dir1" </> "dir2" in
+      ( desc <> ", isDeleted"
+      -- , \fp -> createDirectoryIfMissing True (fp </> p)
+      , const (return ())
+      , removePathForcibly
+      , [("", dirEvent Event.isDeleted)]
+      )
+   ]
+
+-- XXX add metadata change tests for dirs?
+dirTests :: String -> String -> String -> [TestDesc]
+dirTests desc dir1 dir2 =
+    let createDir1 fp = createDirectoryIfMissing True (fp </> dir1)
+    in
+    [
+      ( desc <> ", isCreated"
+      , const (return ())
+      , createDir1
+      , [ (dir1, dirEvent Event.isCreated) ]
+      )
+    , ( desc <> ", isDeleted"
+      , createDir1
+      , \fp -> removePathForcibly (fp </> dir1)
+      , [ (dir1, dirEvent Event.isDeleted) ]
+      )
+    , ( desc <> ", isMoved"
+      , createDir1
+      , \fp -> renameDirectory (fp </> dir1) (fp </> dir2)
+      , [ (dir1, dirEvent Event.isMoved)
+        , (dir2, dirEvent Event.isMoved)
+        ]
+      )
+    ]

```

Comment:
Do we test for events like "root unwatched and root moved"

Reply:
That would go in rootTests if we do. There are more tests that can be added. You can add more once this PR is committed.